pkg/security/x509source.go

```diff
@@ -312,6 +323,115 @@ func (x *x509source) requestFromSentry(ctx context.Context, csrDER []byte) ([]*x
 	return workloadcert, nil
 }
 
+// maybeWriteSVIDToDir writes the current SVID to the configured directory if
+// configured. Does this atomically using unix symlinks.
+func (x *x509source) maybeWriteSVIDToDir() error {
+	if x.writeToDiskDir == nil || x.currentSVID == nil || x.trustAnchors == nil {
+		return nil
+	}
+
+	certs, pk, err := x.currentSVID.Marshal()
+	if err != nil {
+		return err
+	}
+
+	ta, err := x.trustAnchors.Marshal()
+	if err != nil {
+		return err
+	}
+
+	if err := atomicWrite(x.clock, *x.writeToDiskDir, map[string][]byte{
+		"tls.crt": certs,
+		"tls.key": pk,
+		"ca.crt":  ta,
+	}); err != nil {
+		return fmt.Errorf("unable to write SVID to disk: %w", err)
+	}
+
+	log.Infof("Wrote SVID to %q", *x.writeToDiskDir)
+
+	return nil
+}
+
+func (x *x509source) updateTrustAnchorFromFile(ctx context.Context, filepath string) error {
+	x.lock.RLock()
+	defer x.lock.RUnlock()
+
+	rootPEMs, err := os.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf("failed to read trust anchors file %q: %s", filepath, err)
+	}
+
+	trustAnchorCerts, err := secpem.DecodePEMCertificates(rootPEMs)
+	if err != nil {
+		return fmt.Errorf("failed to decode trust anchors: %s", err)
+	}
+
+	x.trustAnchors.SetX509Authorities(trustAnchorCerts)
+
+	if err := x.maybeWriteSVIDToDir(); err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	wg.Add(len(x.trustAnchorSubscribers))
+	for _, ch := range x.trustAnchorSubscribers {
+		go func(chi chan<- struct{}) {
+			defer wg.Done()
+			select {
+			case chi <- struct{}{}:
+			case <-ctx.Done():
+			}
+		}(ch)
+	}
+
+	return nil
+}
+
+func atomicWrite(clock clock.Clock, dir string, data map[string][]byte) error {
+	dir = filepath.Clean(dir)
+	if err := os.MkdirAll(filepath.Dir(dir), 0o700); err != nil {
+		return err
+	}
+
+	// Replace ':' with '-' in the timestamp to avoid issues with Windows
+	// filesystems.
+	newDir := dir + "-" + strings.ReplaceAll(clock.Now().Format(time.RFC3339), ":", "-")
+	if err := os.MkdirAll(newDir, 0o700); err != nil {
```

Comment:
```suggestion
	newDir := dir + "-" + clock.Now().Format("20060102-150405-999999999")
	if err := os.MkdirAll(newDir, 0o700); err != nil {
```

Reply:
Thanks!

Reply:
The nanosecond exponent of this format wasn't working for me, so have left it at millisecond precision. Think this is enough anyway :slightly_smiling_face:

Reply:
Oh yes for sure